sdk/core/azure-core/src/tracing/tracing.cpp

```diff
@@ -30,6 +30,13 @@ namespace Azure { namespace Core { namespace Tracing { namespace _internal {
 
   using Azure::Core::Context;
 
+  std::shared_ptr<TracerProviderImpl> TracerImplFromTracer(
+      std::shared_ptr<TracerProvider> const& provider)
+  {
+    const auto pointer = static_cast<TracerProvider*>(provider.get());
+    return std::shared_ptr<TracerProviderImpl>(provider, pointer);
+  }
```

Comment:
This is `std::shared_ptr` magic stuff, I strongly doubt it has a name because it's not a pattern (or rather the pattern's name is `std::static_pointer_cast<T>`).

I couldn't get `std::static_pointer_cast<T>` to work here, so I pulled the relevant guts out of it and used them.

It relies on the ctor for std::shared_ptr which takes a pointer to the _Ref part of the std::shared_ptr and the _Ptr part of the std::shared_ptr. And yes, this is all part of the design for std::shared_ptr. This uses case 8 of [the std::shared_ptr ctor](https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr). 

And yes, it's not at all obvious - I got this trick from STL, if it matters.